File: src/status/tray-presence.ts
// src/status/tray-presence.ts
//
// Vigila que el icono de bandeja siga vivo en la sesión del usuario (Windows).
//
// ── Por qué hace falta vigilarlo ─────────────────────────────────────
//
//   El MSI cierra la bandeja para poder actualizarla: su .exe es single-file
//   self-contained y reemplazarlo bajo un proceso vivo lo rompe. Pero nadie la
//   vuelve a abrir — la Run key solo dispara en el siguiente inicio de sesión.
//   Así que tras CADA auto-actualización el usuario se queda sin bandeja
//   durante días, hasta que reinicia.
//
//   Eso era un icono ausente. Desde ADR-0012 es otra cosa: la bandeja es donde
//   vive el INDICADOR de "te están viendo la pantalla" y donde se enseña el
//   DIÁLOGO de consentimiento. Sin ella:
//
//     · el indicador no aparece — alguien mira sin que se note;
//     · la petición de consentimiento no la lee nadie, vence, y eso cuenta
//       como negativa: todas las sesiones rechazadas, en silencio.
//
//   O sea, la función de privacidad se apagaría sola después de cada
//   actualización. Por eso esto no es cosmético.
//
// ── Por qué se comprueba periódicamente y no solo al arrancar ────────
//
//   El arranque del agente NO coincide con el momento en que hace falta. El
//   caso típico es: el agente lleva horas corriendo, se actualiza, el MSI mata
//   la bandeja, y el servicio se reinicia DESPUÉS o ni siquiera. Y hay otros
//   caminos —el usuario la cierra, o entra en sesión más tarde que el
//   servicio— en los que un único intento al inicio llega demasiado pronto.
//
//   El coste de mirar es un enumerado de procesos cada pocos minutos, y solo
//   se actúa cuando falta.

import type { AgentContext } from "../core/agent-context";

/**
 * Cada cuánto se comprueba. Cinco minutos: la bandeja no es urgente al
 * segundo, y un intervalo corto solo añadiría ruido en el log de PrivSvc.
 */
export const TRAY_CHECK_INTERVAL_MS = 5 * 60 * 1000;

let timer: NodeJS.Timeout | null = null;

/** Una comprobación. Nunca lanza. */
export async function ensureTrayOnce(ctx: AgentContext): Promise<void> {
  if (process.platform !== "win32") return;
  try {
    const res: any = await (ctx.priv as any)?.call?.({
      v: 1,
      id: `tray.ensure.${Date.now()}`,
      method: "tray.ensure",
      params: {}
    });
    const action = res?.result?.action ?? res?.action;
    // Solo se registra cuando PASA algo. "already_running" es el caso normal y
    // llenaría el log a razón de doce líneas por hora sin decir nada.
    if (action && action !== "already_running" && action !== "no_console_session") {
      ctx.logger?.info?.("[tray] presencia", { action });
    }
  } catch (err: any) {
    // Un PrivSvc que no responde ya se registra en su propio camino; aquí no
    // se puede hacer nada y desde luego no tumbar el agente.
    ctx.logger?.debug?.("[tray] no se pudo comprobar la presencia", {
      err: err?.message || String(err)
    });
  }
}

/** Arranca la vigilancia. Idempotente. */
export function startTrayPresenceWatch(ctx: AgentContext): void {
  if (process.platform !== "win32" || timer) return;
  // Una primera pasada en cuanto haya IPC, y luego el ritmo lento.
  void ensureTrayOnce(ctx);
  timer = setInterval(() => void ensureTrayOnce(ctx), TRAY_CHECK_INTERVAL_MS);
  timer.unref?.();
}

export function stopTrayPresenceWatch(): void {
  if (timer) {
    clearInterval(timer);
    timer = null;
  }
}

File: privsvc/windows/Tracenium.PrivSvc.Windows/Ipc/TrayPresence.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Runtime.InteropServices;

namespace Tracenium.PrivSvc.Windows.Ipc;

/// <summary>
/// Asegura que el icono de bandeja está corriendo en la sesión de consola.
///
/// ── El problema que resuelve ─────────────────────────────────────────
///
///   El MSI TIENE que cerrar la bandeja para actualizarla: su .exe es
///   single-file self-contained y reemplazarlo bajo un proceso vivo lo deja
///   sin poder cargar sus propios assemblies (ver el bloque CloseApplication
///   en Product.wxs, y el incidente que lo motivó).
///
///   Pero nadie la vuelve a abrir. La Run key solo dispara en el siguiente
///   inicio de sesión, así que tras CADA auto-actualización el usuario se
///   queda sin bandeja hasta que reinicia o vuelve a entrar. En campo eso son
///   días.
///
/// ── Por qué ahora importa mucho más que antes ────────────────────────
///
///   La bandeja dejó de ser decorativa. Desde ADR-0012 es donde vive el
///   INDICADOR de "te están viendo la pantalla" y donde se enseña el DIÁLOGO
///   de consentimiento. Sin bandeja corriendo:
///
///     - el indicador no aparece — alguien mira sin que se note;
///     - la petición de consentimiento no la lee nadie, vence, y el agente la
///       trata como negativa: todas las sesiones rechazadas, en silencio.
///
///   O sea: la función de privacidad que acabamos de construir se apaga sola
///   después de cada actualización.
///
/// ── Por qué SÍ es viable, contra lo que decía el WiX ─────────────────
///
///   El comentario del instalador dice que relanzarla desde el MSI no es
///   viable porque el auto-update corre como SYSTEM y un GUI lanzado desde ahí
///   quedaría huérfano en la sesión 0. Eso es cierto para un spawn normal — y
///   es exactamente el problema que SessionScreenCapture ya resuelve para el
///   helper de captura: WTSQueryUserToken + CreateProcessAsUser con
///   lpDesktop="winsta0\\default". La maquinaria existe; solo no se había
///   aplicado aquí.
///
/// ── Token NORMAL, no el elevado ──────────────────────────────────────
///
///   SessionScreenCapture pide el token ENLAZADO para poder inyectar entrada
///   en ventanas elevadas. Aquí NO: un icono de bandeja no necesita
///   privilegios, y lanzarlo elevado sería regalar integridad alta a un
///   proceso que solo lee un JSON y dibuja una ventana.
/// </summary>
internal static class TrayPresence
{
    private const string TrayExeName = "Tracenium.AgentTray.exe";

    /// <summary>
    /// Arranca la bandeja si no está corriendo en la sesión de consola.
    ///
    /// Devuelve qué hizo, para que el llamador lo registre: "already_running"
    /// es el caso normal y no debería llenar el log.
    /// </summary>
    public static PrivSvcResponse Ensure(PrivSvcRequest req)
    {
        try
        {
            var session = SessionScreenCapture.NativeMethods.WTSGetActiveConsoleSessionId();
            if (session == 0xFFFFFFFF)
            {
                // Nadie en consola: no hay bandeja que arrancar y no es un
                // error. Un servidor sin sesión interactiva vive así siempre.
                return PrivSvcResponse.Success(req.Id, new { ok = true, action = "no_console_session" });
            }

            if (IsRunningInSession(session))
            {
                return PrivSvcResponse.Success(req.Id, new { ok = true, action = "already_running" });
            }

            var exe = ResolveTrayExe();
            if (exe is null)
            {
                return PrivSvcResponse.Success(req.Id, new
                {
                    ok = false,
                    action = "exe_not_found",
                    hint = "Tracenium.AgentTray.exe no está donde se esperaba"
                });
            }

            Launch(exe, session);
            IpcLog.Write($"tray relaunched into session {session}: {exe}");
            return PrivSvcResponse.Success(req.Id, new { ok = true, action = "launched" });
        }
        catch (Exception ex)
        {
            // Nunca tumbar por esto: la bandeja es importante y no es crítica
            // para que el agente siga funcionando.
            IpcLog.Write($"tray ensure failed: {ex.Message}");
            return PrivSvcResponse.Success(req.Id, new
            {
                ok = false,
                action = "error",
                hint = ex.Message
            });
        }
    }

    /// <summary>
    /// ¿Hay una bandeja viva EN ESA SESIÓN?
    ///
    /// Comprobar solo el nombre no basta en un equipo con varias sesiones
    /// (RDP, cambio rápido de usuario): la bandeja de otro usuario no le sirve
    /// de nada a quien está en consola.
    /// </summary>
    private static bool IsRunningInSession(uint session)
    {
        var name = Path.GetFileNameWithoutExtension(TrayExeName);
        foreach (var p in Process.GetProcessesByName(name))
        {
            try
            {
                if ((uint)p.SessionId == session) return true;
            }
            catch
            {
                // El proceso murió entre el enumerate y el acceso.
            }
            finally
            {
                p.Dispose();
            }
        }
        return false;
    }

    /// <summary>
    /// Ruta del .exe de la bandeja.
    ///
    /// PrivSvc vive en ...\Tracenium\PrivSvc\ y la bandeja en
    /// ...\Tracenium\AgentTray\. Se busca relativo al propio binario y NO por
    /// una ruta absoluta escrita a mano: el instalador puede ir a Program
    /// Files (x86) en un equipo de 32 bits, y una constante allí fallaría solo
    /// en esos.
    /// </summary>
    private static string? ResolveTrayExe()
    {
        var baseDir = AppContext.BaseDirectory;
        string[] candidates =
        {
            Path.Combine(baseDir, TrayExeName),
            Path.Combine(baseDir, "..", "AgentTray", TrayExeName),
            Path.Combine(baseDir, "..", TrayExeName)
        };
        foreach (var c in candidates)
        {
            var full = Path.GetFullPath(c);
            if (File.Exists(full)) return full;
        }
        return null;
    }

    private static void Launch(string exe, uint session)
    {
        if (!SessionScreenCapture.NativeMethods.WTSQueryUserToken(session, out var userToken))
        {
            throw new InvalidOperationException(
                $"WTSQueryUserToken failed (Win32 {Marshal.GetLastWin32Error()}).");
        }

        var primaryToken = IntPtr.Zero;
        var envBlock = IntPtr.Zero;
        try
        {
            // ⚠️ El token TAL CUAL, sin pedir el enlazado. Ver la nota de
            // cabecera: la bandeja no necesita elevación y dársela sería
            // ampliar su superficie a cambio de nada.
            if (!SessionScreenCapture.NativeMethods.DuplicateTokenEx(userToken,
                    SessionScreenCapture.NativeMethods.TOKEN_ALL_ACCESS, IntPtr.Zero,
                    SessionScreenCapture.NativeMethods.SecurityImpersonation,
                    SessionScreenCapture.NativeMethods.TokenPrimary, out primaryToken))
            {
                throw new InvalidOperationException(
                    $"DuplicateTokenEx failed (Win32 {Marshal.GetLastWin32Error()}).");
            }

            SessionScreenCapture.NativeMethods.CreateEnvironmentBlock(out envBlock, primaryToken, false);

            var si = new SessionScreenCapture.NativeMethods.STARTUPINFO();
            si.cb = Marshal.SizeOf<SessionScreenCapture.NativeMethods.STARTUPINFO>();
            // Sin esto el proceso arranca en un escritorio que no existe para
            // el usuario y el icono nunca aparece — el "huérfano en la sesión
            // 0" que el WiX describía.
            si.lpDesktop = "winsta0\\default";

            var created = SessionScreenCapture.NativeMethods.CreateProcessAsUser(
                primaryToken,
                exe,
                null,
                IntPtr.Zero,
                IntPtr.Zero,
                false,
                SessionScreenCapture.NativeMethods.CREATE_UNICODE_ENVIRONMENT,
                envBlock,
                Path.GetDirectoryName(exe),
                ref si,
                out var pi);

            if (!created)
            {
                throw new InvalidOperationException(
                    $"CreateProcessAsUser failed (Win32 {Marshal.GetLastWin32Error()}).");
            }

            // No se espera al proceso: la bandeja vive mientras dure la sesión
            // del usuario. Solo se sueltan los handles.
            if (pi.hProcess != IntPtr.Zero) SessionScreenCapture.NativeMethods.CloseHandle(pi.hProcess);
            if (pi.hThread != IntPtr.Zero) SessionScreenCapture.NativeMethods.CloseHandle(pi.hThread);
        }
        finally
        {
            if (envBlock != IntPtr.Zero) SessionScreenCapture.NativeMethods.DestroyEnvironmentBlock(envBlock);
            if (primaryToken != IntPtr.Zero) SessionScreenCapture.NativeMethods.CloseHandle(primaryToken);
            SessionScreenCapture.NativeMethods.CloseHandle(userToken);
        }
    }
}
